applications/FluidDynamicsApplication/custom_elements/symbolic_stokes.cpp

```diff
@@ -198,6 +198,85 @@ void SymbolicStokes<TElementData>::AddBoundaryTraction(
+        }
+        data.N = data_N;
+        // Shape functions gradients
+        const GeometryData::IntegrationMethod integration_method = GeometryData::GI_GAUSS_2;
+        ShapeFunctionDerivativesArrayType shape_derivatives;
+        Vector DetJ;
+        r_geometry.ShapeFunctionsIntegrationPointsGradients(shape_derivatives, DetJ, integration_method);
+        data.DN_DX = shape_derivatives[0]; // Note: Valid only for linear elements
+        // Compute strain and stress
+        this->CalculateMaterialResponse(data);
+        // Add limits to the strain, and recompute stress if required
+        const auto &constitutive_law = this->GetConstitutiveLaw();
+        auto strain = data.StrainRate;
+        const double gamma_dot = std::sqrt(2.0 * (strain[0] * strain[0] + strain[1] * strain[1] + strain[2] * strain[2]) + strain[3] * strain[3] + strain[4] * strain[4] + strain[5] * strain[5]);
+        double gamma_dot_from_constitutive_law = constitutive_law->GetValue(EQ_STRAIN_RATE, gamma_dot_from_constitutive_law);
+        double ratio = fabs(gamma_dot_from_constitutive_law / gamma_dot);
+        if (ratio < 0.99999 && ratio > 1e-20) {
+            strain *= ratio;
+            auto &r_values = data.ConstitutiveLawValues;
+            Flags &ConstitutiveLawOptions = r_values.GetOptions();
+            ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_STRESS);
+            ConstitutiveLawOptions.Set(ConstitutiveLaw::COMPUTE_CONSTITUTIVE_TENSOR, false);
+            r_values.SetShapeFunctionsValues(data_N);
+            r_values.SetStrainVector(strain);
+            Vector stress(strain.size());
+            r_values.SetStressVector(stress); //this is an ouput parameter
+            constitutive_law->CalculateMaterialResponseCauchy(r_values);
+            rOutput = inner_prod(stress, strain);
+        }
+        else {
+            rOutput = inner_prod(data.ShearStress, data.StrainRate);
+        }
+    }
+    else if (rVariable == EQ_STRAIN_RATE || rVariable == EFFECTIVE_VISCOSITY) {
+        TElementData data;
+        data.Initialize(*this, rCurrentProcessInfo);
+        rOutput = 0.0;
+        // Shape functions
+        const GeometryType &r_geometry = this->GetGeometry();
+        const unsigned int num_nodes = r_geometry.PointsNumber();
+        Vector data_N(num_nodes);
+        for (unsigned int i = 0; i < num_nodes; i++) {
+            data_N[i] = 1.0 / (double)num_nodes;
+        }
+        data.N = data_N;
+        // Shape function gradients
+        const GeometryData::IntegrationMethod integration_method = GeometryData::GI_GAUSS_2;
+        ShapeFunctionDerivativesArrayType shape_derivatives;
+        Vector DetJ;
+        r_geometry.ShapeFunctionsIntegrationPointsGradients(shape_derivatives, DetJ, integration_method);
+        data.DN_DX = shape_derivatives[0]; // Note: Valid only for linear elements
+        // Compute strain and stress
+        this->CalculateMaterialResponse(data);

```

Comment:
I assume that you calculate the material response again in case the effective viscosity is computed as the ratio of stress - strain norms. Am I wrong?

Reply:
That is correct. Apart from this constitutive model can use some shear-thinning and temperature dependency models to compute the effective viscosity. In either case, we need to compute the material response.

Reply:
Agree. In all these cases we're obliged to calculate the material response. :+1: